### PR TITLE
- fix Windows shared build

### LINKED ISSUE #639: Windows shared builds are failing
Tested both with 1.14.1 and master, same result.

cmake .. -A x64 -DBUILD_SHARED_LIBS=On && cmake --build .

> (Link target) ->
 LINK : fatal error LNK1104: cannot open file 'Debug\benchmark.lib'

### PR DESCRIPTION
**Pull request by SSE4**
fixed #639 
related to the #635
/cc @Croydon @p-groarke

### CODE REVIEW

**Comment by SSE4** [comment]
I signed it!

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -22,10 +22,16 @@ set_target_properties(benchmark PROPERTIES
   VERSION ${GENERIC_LIB_VERSION}
   SOVERSION ${GENERIC_LIB_SOVERSION}
 )
+
+include(GenerateExportHeader)
+generate_export_header(benchmark EXPORT_FILE_NAME ${CMAKE_CURRENT_SOURCE_DIR}/../include/benchmark_export.h)
```
Comment (LebedevRI) [inline src/CMakeLists.txt]: It's a very bad idea to write outside of the `CMAKE_CURRENT_BINARY_DIR`

Comment (SSE4) [reply src/CMakeLists.txt]: thanks, I'll try to change to use CMAKE_CURRENT_BINARY_DIR. initial attempt produced some CMake errors for me. also, build fails hard on tests, I'll take a look.

**Comment by SSE4** [comment]
it seems to be dual build-system project (Bazel + CMake), so CMake's **generate_export_header** doesn't seem to be good fit here (I don't want to re-implement the same in Bazel, especially as I have no experience with mentioned build system). probably, the best thing I can do for now is to just copy generated export definitions into **benchmark/benchmark.h**, does it sound like an acceptable solution?

**Comment by LebedevRI** [comment]
I would suggest looking how googletest handles this issue.
They too have `gtest_main`.

**Comment by SSE4** [comment]
they do not use **generate_export_header** and have their own macro definitions (**GTEST_API_**) :
https://github.com/google/googletest/blob/41f0e24335d8874cc87420a6349bb1844e2a15de/googletest/include/gtest/internal/gtest-port.h#L969

**Comment by LebedevRI** [comment]
I must say that my personal opinion is that this is kinda ugly.
I wonder if it would be better to simply do
```
if(BUILD_SHARED_LIBS AND WIN32)
  message(SEND_ERROR "Building libbenchmark as shared library is not supported")
endif()
```

**Comment by dominichamon** [comment]
How does this work on installation on linux when the public header is copied to /usr and benchmark_export.h doesn't exist? or does it also get created for linux, and get copied to /usr on installation?

This seems awfully intrusive. I think short term not supporting shared libs on windows is reasonable, until we can figure out a better way.

**Comment by SSE4** [comment]
I've asked @Croydon and he is okay with not supporting Windows shared built at all, works for me as well.

**Comment by dominichamon** [comment]
Can this PR be closed then?

**Comment by SSE4** [comment]
yes

**Pull request closed without merging**